VXLAN UDP Port configuration support

selansen opened the pull request:
This PR chnages allow user to configure VxLAN UDP
    port number. By default we use 4789 port number. But this commit
    will allow user to configure port number during swarm init.
    VxLAN port can't be modified after swarm init.

Signed-off-by: selansen <elango.siva@docker.com>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

AkihiroSuda commented on docs/api/version-history.md:
```diff
@@ -12,6 +12,10 @@ keywords: "API, Docker, rcli, REST, documentation"
+## V1.40 API changes
+
+[Docker Engine API v1.39](https://docs.docker.com/engine/api/v1.40/) documentation
+* `POST /swarm/init` now accepts a `vxlan-udp-port` property to set vxlan udp port number.
```
Needs to be moved to L34

thaJeztah commented on api/swagger.yaml:
```diff
@@ -8876,13 +8876,19 @@ paths:
                   SubnetSize specifies the subnet size of the networks created from the default subnet pool
                 type: "integer"
                 format: "uint32"
+              VxlanUDPPort:
```
`VxlanUDPPort` is a bit verbose here, and very much tied to the implementation; perhaps some other naming would be better.

*However* - if I'm looking at this (and understand correctly), this configures the port to used for the overlay data path. We already have `DataPathAddr`, which defines the address to use. `ListenAddr` and `AdvertiseAddr` both accept an ip/interface and optionally a port. Why don't we use the same approach for this one? So;

Change `DataPathAddr` to accept the format: `<ip|interface>[:<port>]`

- if only `<ip|interface>` is specified, use the default port (4789)
- if both `<ip|interface>` and `<port>` are specified, use the given address and port
- if none is specified, use `ListenAddr` as address, and `4789` as port.

That change would be backward-compatible, because the port is optional


I _think_ this approach actually came up in a discussion when the `DataPathAddr` was added (https://github.com/moby/moby/pull/32717) I can't find that discussion though, so possibly it was offline (or on Slack). @fcrisciani may recall

selansen replied:
We would like to use single VXLAN port across all the node in the same cluster.  Datapathaddr can be configured via swarm init and  as well swarm join.  Issue here is user will have to remember and pass same port number for each swarm join. Thats going to be error prone. We kind of discussed this option and moved away from it.
I wanted to make sure we discuss  design related issues before starting feature development  and thats the main reason I had written proposal https://github.com/moby/moby/issues/38030 ( before working on this feature )to make sure we discuss any design specific issue before implementation. 
Pls let me know if you still have concern with this approach. 

Regrading VxlanUDPPort name : I wanted to convey this port is used only for vxlanport ( on UDP). Initial name was suggested like overlayvxlanport or something like that. But I thought this is simple :). Let me know, I can try to come up with some other name.

thaJeztah replied:
> Datapathaddr can be configured via swarm init and as well swarm join. Issue here is user will have to remember and pass same port number for each swarm join. Thats going to be error prone.

Let me try to understand; 

#### Old situation (current release)

- `docker swarm init` initializes the swarm, and uses port `4789` (hard-coded)
- `docker swarm join` joins a node to the swarm, and also uses port `4789` (hard-coded)


### Current flow in this PR

- Run `docker swarm init --vxlan-udp-port=1234` to initialize it as a manager
- On another node, run `docker swarm join <manager-ip>` to join the node to the swarm
- The node somehow gets the port number from the manager, and sets `vxlan-udp-port` to the same port as the manager

or

- On another node, run `docker swarm join --vxlan-udp-port=4567 <manager-ip>` to join the node to the swarm, but using a different port-number than is used by the manager


Question:

- How is the port communicated from the manager to the node that is joining? Or will this be printed as part of the ouput of `docker swarm init`? ("To add a worker to this swarm, run the following command:")? 
- I wonder if we can make the assumption that all nodes in the swarm are expected to be using the same port (I can foresee situations where a different port number is needed for _some_ nodes and/or data-centers, but possibly not for _all_ the nodes)

### Alternative proposal

So, thinking about my proposal;

- Run `docker swarm init --data-path-addr=1.2.3.4:4321` to initialize a swarm
- Run `docker swarm join` to join a node
  - should this use the default (`4789`) port?
  - or, should this use the same port as the manager uses (`4321`)?
  - note that the manager can still communicate the port-number to the node (as it _does_ have its portnumber set)
- Run `docker swarm join --data-path-addr=2.3.4.5:1234` to join the node to the swarm, but using a different port-number than is used by the manager

> Or will this be printed as part of the ouput of `docker swarm init`? ("To add a worker to this swarm, run the following command:")? 

If that is the case, we can consider making the `<ip|interface>` optional;

`--data-path-addr` | data-path-address | vxlan port
---|---|---
(not set) | `ListenAddr` (default) | `4789` (default)
`1.2.3.4` | `1.2.3.4` |  `4789` (default)
`eth1` | IP-addres of `eth1` |  `4789` (default)
`1.2.3.4:4321` | `1.2.3.4` |  `4321`
`eth1:4321` | IP-addres of `eth1` |  `4321`
`:4321` | `ListenAddr` (default) |  `4321`


The above is the approach that's used for the `-H` option of `dockerd`;

```
dockerd -H 172.18.0.1
INFO[2018-11-09T15:03:37.583248272Z] API listen on 172.18.0.1:2375 

dockerd -H 172.18.0.1:4321
INFO[2018-11-09T15:03:37.583248272Z] API listen on 172.18.0.1:4321 

dockerd -H :4321
INFO[2018-11-09T15:00:19.352779504Z] API listen on 127.0.0.1:4321   
```

### Naming

> Regrading VxlanUDPPort name : I wanted to convey this port is used only for vxlanport ( on UDP). Initial name was suggested like overlayvxlanport or something like that. But I thought this is simple :). Let me know, I can try to come up with some other name.

I think that's something to be communicated through the _documentation_ (both for the CLI, and in the description of the Swagger) and (if room permits) the flag description. 

For the flag description, this could be something like one of these;

```
Port to use for data path traffic (UDP)
Port (UDP) to use for data path traffic
UDP port to use for data path traffic
```

Given that this relates to the data-path,  my preference would be to name both the same / similar:

- `DataPathPort` for the API
- `--data-path-port` for the CLI

That way, it's clear what this option does, without having to understand what VxLAN is used for (and/or leaking that implementation detail into the UX). It will also show both options together in the CLI usage output;

```
Usage:	docker swarm join [OPTIONS] HOST:PORT

Join a swarm as a node and/or manager

Options:
      --advertise-addr string   Advertised address (format: <ip|interface>[:port])
      --availability string     Availability of the node ("active"|"pause"|"drain") (default "active")
      --data-path-addr string   Address or interface to use for data path traffic (format: <ip|interface>)
      --data-path-port int      Port to use for data path traffic (UDP) (default 4789)
      --listen-addr node-addr   Listen address (format: <ip|interface>[:port]) (default 0.0.0.0:2377)
      --token string            Token for entry into the swarm
```

selansen replied:
I had offline discussion with @thaJeztah and summary is below:

data-path-addr is configured per node ( manager or worker). So we added option in swarm init & swarm join.
vxlanport - is configured per cluster. ( Main  use case for this feature  is what if port "4789"  is blocked.  In general customers prefer to stick to single port because if they use different port for each node, then trying to open/expose  port for any service - they need to remember all the ports that have been used in each and every node in the cluster. It will create unnecessary headache  )

vxlan port number will be communicated via node object information from leader to worker node after they join the cluster.

 Naming:
=======

I still prefer to keep it vxlan-udp-port at the end of the day this what it represents.   But I can change it to --data-path-port instead of vxlanport.
Will go ahead and modify the name.

thaJeztah commented on api/types/swarm/swarm.go:
```diff
@@ -159,6 +160,7 @@ type InitRequest struct {
 	Availability     NodeAvailability
 	DefaultAddrPool  []string
 	SubnetSize       uint32
+	DataPathPort     uint32
```
We need to fence this feature based on API version; when using an older version of the API, but the daemon "supports" this feature, it should still be ignored; something like this in `initCluster()`;

```go
	version := httputils.VersionFromContext(ctx)
	// DataPathPort was added in API 1.40. Ignore this option on older API versions.
	if versions.LessThan(version, "1.40") {
		req.DataPathPort = 0
	}
```

Here's a patch you can apply:

```patch
diff --git a/api/server/router/swarm/cluster_routes.go b/api/server/router/swarm/cluster_routes.go
index dda2a92d91..337cfd3419 100644
--- a/api/server/router/swarm/cluster_routes.go
+++ b/api/server/router/swarm/cluster_routes.go
@@ -27,6 +27,11 @@ func (sr *swarmRouter) initCluster(ctx context.Context, w http.ResponseWriter, r
                }
                return errdefs.InvalidParameter(err)
        }
+       version := httputils.VersionFromContext(ctx)
+       // DataPathPort was added in API 1.40. Ignore this option on older API versions.
+       if versions.LessThan(version, "1.40") {
+               req.DataPathPort = 0
+       }
        nodeID, err := sr.backend.Init(req)
        if err != nil {
                logrus.Errorf("Error initializing swarm: %v", err)
```

thaJeztah replied:
Sorry, forgot to mention that earlier; can you move this below `DataPathAddr`, so that they're next to each other?

thaJeztah commented on daemon/cluster/listen_addr.go:
```diff
@@ -123,6 +123,22 @@ func validateDefaultAddrPool(defaultAddrPool []string, size uint32) error {
+	// The Dynamic Ports, aka the Private Ports, from 49152-65535
+	// So we can allow range between 1024 to 49151
+	if portNum < 1024 || portNum > 49151 {
+		return fmt.Errorf("validateDataPathPort : Data Path Port number is not in valid range %d", portNum)
```
Can you;

- remove the `validateDataPathPort:` here? The error message should only describe the error condition (not what function did the validation; it's not a debug-message).
- `Path` and `Port` should be lowercase (and errors should not start with an uppercase)
- Add the _valid_ range in the error-message (this will help the user resolve  the problem if they specify an invalid number).

Perhaps;

```suggestion
		return fmt.Errorf("data path port number is not in valid range (1024-49151): %d", portNum)
```

thaJeztah commented on daemon/cluster/listen_addr.go:
```diff
@@ -123,6 +123,22 @@ func validateDefaultAddrPool(defaultAddrPool []string, size uint32) error {
 	return nil
 }
 
+// validateDataPathPort validates vxlan udp port (data path port) number.
```
Might want to add that;

- valid port numbers are between `1024` and `49151`
- if no port is set, the default (`4789`) is returned

selansen replied:
I already added those two comments inside the function.  Thought it will be redundant.
I can add it in API comment section too.

thaJeztah commented on daemon/cluster/listen_addr.go:
```diff
@@ -123,6 +123,22 @@ func validateDefaultAddrPool(defaultAddrPool []string, size uint32) error {
+func validateDataPathPort(portNum uint32) error {
+	// if the value comes as 0 by any reason we set it to default value 4789
+	if portNum == 0 {
+		portNum = 4789
```
So; does this actually work? Here's a quick test; (you can run it here: https://play.golang.org/p/MnC5gxYX26F)

```go
package main

import (
	"fmt"
)

type InitRequest struct {
	DataPathPort uint32
}

func main() {
	req := InitRequest{}
	if err := validateDataPathPort(req.DataPathPort); err != nil {
		fmt.Println(err)
	}
	fmt.Printf("Data port number: %d", req.DataPathPort)
}

func validateDataPathPort(portNum uint32) error {
	// if the value comes as 0 by any reason we set it to default value 4789
	if portNum == 0 {
		portNum = 4789
	}
	// IANA procedures for each range in detail
	// The Well Known Ports, aka the System Ports, from 0-1023
	// The Registered Ports, aka the User Ports, from 1024-49151
	// The Dynamic Ports, aka the Private Ports, from 49152-65535
	// So we can allow range between 1024 to 49151
	if portNum < 1024 || portNum > 49151 {
		return fmt.Errorf("validateDataPathPort : Data Path Port number is not in valid range %d", portNum)
	}
	return nil
}
```

I'm not always a big fan of validation-functions modifying the value; I see we do this in other places, but perhaps for this one we should change the signature to something like;

```go
func getDataPathPort(portNum uint32) (uint32, error) {
    // ....
}
```

Or pass the whole `req` (as is done in `validateAndSanitizeInitRequest`);

```go
func validateDataPathPort(req *types.InitRequest) error {
    // ....
}
```

thaJeztah replied:
Actually; looks like we're still setting the default in libnetwork as well. Would it be better to just pass `0`, and let libnetwork set the default?

selansen replied:
by looking at other validation/resolve/get APIs, I feel we can stick to your below suggestion. I will change it accordingly.
func getDataPathPort(portNum uint32) (uint32, error) {
    // ....
}

selansen replied:
> Actually; looks like we're still setting the default in libnetwork as well. Would it be better to just pass `0`, and let libnetwork set the default?

This value gets stored into swarm too. So I better change it like I proposed in the above reply.

thaJeztah commented on vendor/github.com/docker/libnetwork/drivers/overlay/overlayutils/utils.go:
```diff
@@ -0,0 +1,44 @@
+	mutex        sync.Mutex
+)
+
+func init() {
```
Can you create a constant for 4789?

```suggestion
const defaultVXLANUDPPort = 4789
func init() {
 	vxlanUDPPort = defaultVXLANUDPPort
}
```

thaJeztah commented on vendor/github.com/docker/libnetwork/drivers/overlay/overlayutils/utils.go:
```diff
@@ -0,0 +1,44 @@
+	defer mutex.Unlock()
+	// if the value comes as 0 by any reason we set it to default value 4789
+	if vxlanPort == 0 {
+		vxlanPort = 4789
```
See my comment above about "where to set the default"

Also; please create a constant for `4789` (see above)

thaJeztah commented on vendor/github.com/docker/libnetwork/drivers/overlay/overlayutils/utils.go:
```diff
@@ -0,0 +1,44 @@
+}
+
+// GetVxlanUDPPort returns Vxlan UDP port number
+func GetVxlanUDPPort() uint32 {
```
`Get` prefixes are [generally not used in Go](https://golang.org/doc/effective_go.html#Getters)

selansen replied:
Got it. But there are other getter functions in the above format so I kept it same. Let me know if this is blocker, else I prefer to keep it the same way.

thaJeztah replied:
It's not a blocker for me; my eye just fell on this one

thaJeztah commented on vendor/github.com/docker/libnetwork/drivers/overlay/overlayutils/utils.go:
```diff
@@ -0,0 +1,44 @@
+}
+
+// ConfigVxlanUDPPort configures vxlan udp port number.
+func ConfigVxlanUDPPort(vxlanPort uint32) error {
```
Note that capitalization doesn't look correct (`VXLAN` is all caps, so this should probably be `ConfigVXLANUDPPORT`)

selansen replied:
I felt "ConfigVxlanUDPPort" is more readable than ConfigVXLANUDPPORT. If you are still not ok with it, I can change it.  pls let me know.

thaJeztah replied:
> I felt "ConfigVxlanUDPPort" is more readable than ConfigVXLANUDPPORT. If you are still not ok with it, I can change it. pls let me know.

Yes, I definitely agree with you that `ConfigVXLANUDPPort` is ugly (due to both `VXLAN` and `UDP` being an accronym, so both are all-caps).

The reason I still would like to see the name changed (also in other locations), is that sticking to a convention makes the code more consistent, and removes "personal preference" for these (naming is already hard! :sweat_smile:).

The danger of "personal preference" (it "looks" better), is that it may trickle up into public API's; in this case; exported functions, but also the actual API.

For example, in our API we have;

- `CpuShares` in `/containers/<id>/json`, but `CPUShares` in `/info`,
- `NetworkSettings.IPAddress`, but [`NetworkSettings.Ports.HostIp`](https://github.com/moby/moby/blob/master/vendor/github.com/docker/go-connections/nat/nat.go#L16-L22)

Changing a public API / exported function once it's released, can be difficult, so ideally, they should be updated soon.

selansen replied:
Ok. Let me change it. I am seeing this comment only now.  Didn't see it 30 mins before.  Not sure if this github issue.

thaJeztah commented on api/swagger.yaml:
```diff
@@ -8876,13 +8876,19 @@ paths:
                 format: "uint32"
+              DataPathPort:
+                description: |
+                  DataPathPort specifies the vxlan udp port number for vxlan traffic
```
- s/vxlan/VXLAN/
- s/udp/UDP/
- Perhaps `data traffic` instead of `VXLAN traffic`


Also, can you

- describe the accepted port range (1024-...)
- describe that if no port is set, or is set to `0`, the default (4789) is used?

thaJeztah commented:
Left some comments; be careful, because GitHub is hiding some of my review comments

selansen commented on integration/network/service_test.go:
```diff
@@ -321,6 +321,48 @@ func noServices(client client.ServiceAPIClient) func(log poll.LogT) poll.Result
 }
 
+func TestServiceWithDatapathPortInit(t *testing.T) {
+	skip.If(t, testEnv.OSType == "windows")
```
good idea.   will do it.

selansen commented:
@thaJeztah , I thought we decided to bump up version. Version changes are little confusing to me. Would like to talk to offline and take it from there.

selansen commented on docs/api/version-history.md:
```diff
@@ -13,6 +13,12 @@ keywords: "API, Docker, rcli, REST, documentation"
+
+[Docker Engine API v1.41](https://docs.docker.com/engine/api/v1.41/) documentation
+
+* `POST /swarm/init` now accepts a `data-path-port` property to set data path port number.
```
below is the offline discussion from @thaJeztah 
```For now, it should be in API 1.40
Once we have final approval to get it into the 18.09.1 release, we need to do a follow up to;

- bump the API version to 1.41
- move all features except for yours to 1.41
- cherry pick your changes```

thaJeztah commented on api/swagger.yaml:
```diff
@@ -8876,13 +8876,21 @@ paths:
+                  DataPathPort specifies the data path port number for data traffic.
+                  Acceptable port range is 1024 to 49151.
+                  if no port is set or os set to 0, default port 4789 will be used.
+                type: "integer"
```
Actually, looks like swagger allows specifying a minimum and maximum https://swagger.io/docs/specification/data-models/data-types/

Not sure if that works well for us, because `0` is also valid, so lets keep that for a follow-up to try

thaJeztah commented on daemon/cluster/convert/swarm.go:
```diff
@@ -42,6 +42,7 @@ func SwarmFromGRPC(c swarmapi.Cluster) types.Swarm {
 			RootRotationInProgress: c.RootCA.RootRotation != nil,
 			DefaultAddrPool:        c.DefaultAddressPool,
 			SubnetSize:             c.SubnetSize,
+			DataPathPort:           c.VxlanUDPPort,
```
If we can still rename this (in a follow-up) to `VXLANUDPPort`, that would probably be a good thing to do as well

selansen replied:
For exposing APIs and names for this feature as datapath port is good. But at the end of the day internally its going to be VXLAN  UDP port. Thats why I wanted to retain the same name at least inside swarmkit & libnetwork. Again, let me know if you think its still not a valid reason.

thaJeztah replied:
Sorry, I meant `VXLANUDPPort` (capitalisation), to be consistent with the rest (and libnetwork)

selansen replied:
got it .

thaJeztah replied:
hm... looks like we missed this one in the swarmkit change; I opened a PR; https://github.com/docker/swarmkit/pull/2783

thaJeztah replied:
Let me see if we can get https://github.com/docker/swarmkit/pull/2783 merged 
(https://github.com/moby/moby/pull/38102/files#r234425632)

thaJeztah commented on integration/network/service_test.go:
```diff
@@ -321,6 +321,49 @@ func noServices(client client.ServiceAPIClient) func(log poll.LogT) poll.Result
+	// Clean up , set it back to original one to make sure other tests don't fail
+	datapathPort = 4789
+	ops = append(ops, daemon.WithSwarmDataPathPort(datapathPort))
+	d = swarm.NewSwarm(t, testEnv, ops...)
```
I'm not sure why this would be needed; `swarm.NewSwarm()` creates a _new_ daemon, so other tests should not be affected?

Oh..hm... perhaps in the "remote daemon" situation, where it doesn't create a new daemon; https://github.com/moby/moby/blob/d3cc071bb98669545d4e3043c9bd85879292b815/integration/internal/swarm/service.go#L50-L61

selansen replied:
when I wrote test cases for local scope default address pool feature( where we change default IP address ), few of the old test cases were failing. My initial assumption was each time we spawn new daemon ,  everything is fresh. Following  test  cases should not affect. But I was wrong. so I added clean up code( setting back to default value) after I complete my testing .  I continue the same practice in all new features that I am involved.

thaJeztah replied:
If this has to be done for all tests, perhaps we should have a utility for it, and make it part of the `defer setupTest(t)()`

(not needed to do that in this PR)

thaJeztah commented on integration/network/service_test.go:
```diff
@@ -321,6 +321,49 @@ func noServices(client client.ServiceAPIClient) func(log poll.LogT) poll.Result
+	d.Stop(t)
+
+	// Clean up , set it back to original one to make sure other tests don't fail
+	datapathPort = 4789
```
Perhaps we should test _without_ setting to the default, and verify that (if no port is specified), the swarm is initialised with the default (`4789`)

selansen replied:
I can try that without setting default.

selansen replied:
I added more logic to test default case too.  In that way it will be doing cleanup and test case for default case too.

thaJeztah commented on integration/network/service_test.go:
```diff
@@ -321,6 +321,49 @@ func noServices(client client.ServiceAPIClient) func(log poll.LogT) poll.Result
 	}
 }
 
+func TestServiceWithDatapathPortInit(t *testing.T) {
```
nit: `TestServiceWithDataPathPortInit` (capital `Path`)

thaJeztah commented on internal/test/daemon/daemon.go:
```diff
@@ -76,6 +76,7 @@ type Daemon struct {
 	SwarmPort       int // FIXME(vdemeester) should probably not be exported
 	DefaultAddrPool []string
 	SubnetSize      uint32
+	DatapathPort    uint32
```
Can you rename this to `DataPathPort` (capital `Path`) to be consistent with the names elsewhere?

thaJeztah reviewed (changes requested):
Left some comments inline

thaJeztah commented on api/swagger.yaml:
```diff
@@ -8876,13 +8876,21 @@ paths:
                   SubnetSize specifies the subnet size of the networks created from the default subnet pool
                 type: "integer"
                 format: "uint32"
+              DataPathPort:
```
Can you move this below `DataPathAddr` (as they "belong" together)?

thaJeztah commented on api/types/swarm/swarm.go:
```diff
@@ -14,6 +14,7 @@ type ClusterInfo struct {
 	RootRotationInProgress bool
 	DefaultAddrPool        []string
 	SubnetSize             uint32
+	DataPathPort           uint32
```
Looks like this field was not added to the swagger.yaml

thaJeztah commented on integration/network/service_test.go:
```diff
@@ -321,6 +321,70 @@ func noServices(client client.ServiceAPIClient) func(log poll.LogT) poll.Result
+
+	var instances uint64 = 1
+	serviceName := "TestService" + t.Name()
+	serviceID := swarm.CreateService(t, d,
```
I'm actually trying to understand this test; where do we test if the data-port is actually set? Why do we start services here (it looks to be only testing that we can create  a service, but I don't see how this would show that the new feature works).

Should we (instead?) check if the `/swarm` endpoint returns the correct port-number?


We'd also need a test to check if a node that joins the swarm correctly copies the port number (but that would probably have to be a test in the e2e)

selansen replied:
when we create overlay (VXLAN)  network , we create VXLAN interface which uses dest port as one of input param. But in our swarm, vxlan network gets created only when we create a service using  overlay network ( lazy creation). So I tried to create service and make sure it succeeds. If the port is in use already , then vxlan interface creation will fail and eventually service create also will fail.

selansen replied:
also checking swarm endpoint would return the value what we stored in RAFT cluster object. we need to make sure we create service with port change to make sure it works at libnetwork level.

thaJeztah replied:
Right, but currently it's more a "regression test"; we're testing that with this feature we're still able to create services, but we don't test;

- if the port we specified is actually stored in RAFT (which we should test by checking the `/swarm` endpoint)
- related to the above; if the feature doesn't work correctly, and (e,g,) we initialize with port 1234, but swarm ignores this, and uses the default port (4789), the test still passes 
- if the service / network is actually using the port we specified (and what's stored in RAFT)


For the last one, not sure how to test; a really dirty way would be to;

- init swarm with a port (eg 5678)
- create a container that publishes port 5678
- create a service

the service should fail to be deployed, because port 5678 is already in use by the container (but we would likely have to check the error to see if that's the reason it failed, and not any other reason)

selansen replied:
We have checks in every module  which  will return error if the port configuration fails during swarm init.  if swarn init with  datapath port succeeds, then we have successfully configured all the way in libnetwork.
To answer your point "we initialize 1234 but swarm takes 4789" - I think we can write unit test in swarmkit RAFT read/write and calling libnetwork code path to make sure swarmkit is working properly. 
For the last one : That doesn't look like the correct way of testing. 

But I will see if I can expand the current test.

thaJeztah replied:
> We have checks in every module which will return error if the port configuration fails during swarm init. if swarn init with datapath port succeeds, then we have successfully configured all the way in libnetwork.

We're currently only testing the "happy path", so we didn't verify if those checks do their job (e.g., there's no test trying to initialise with port `1023` or `49152` - which are "out of range", nor do we have a test for a situation where the specified port overlaps with a port that's already in use) 

> To answer your point "we initialize 1234 but swarm takes 4789" - I think we can write unit test in swarmkit RAFT read/write and calling libnetwork code path to make sure swarmkit is working properly.

unit tests in swarmkit and libnetwork are always good to have (if it's calling libnetwork code, it wouldn't be a unit test though 🤔)

_but_ that would not cover the changes in this PR; given that we process the value in this PR (`getDataPathPort()` can adjust the port if it sees it's `0`); we still need a test here that checks we get the correct port.

> For the last one : That doesn't look like the correct way of testing.

Agreed; so we may need another way to verify it's using the correct port

selansen commented on docs/api/version-history.md:
```diff
@@ -29,6 +29,7 @@ keywords: "API, Docker, rcli, REST, documentation"
   to return those without the specified labels.
 * `POST /containers/create`, `GET /containers/{id}/json`, and `GET /containers/json` now supports
   `BindOptions.NonRecursive`.
+* `POST /swarm/init` now accepts a `data-path-port` property to set data path port number.
```
In that case , I didnt add GET/swarm GET/info for default address pool feature which we delivered as part of 18.09.
Should . I  add for them now ?

thaJeztah replied:
Can you do that in a separate PR?

thaJeztah commented on integration/network/service_test.go:
```diff
@@ -321,6 +321,70 @@ func noServices(client client.ServiceAPIClient) func(log poll.LogT) poll.Result
+
+	poll.WaitOn(t, serviceRunningCount(cli, serviceID, instances), swarm.ServicePoll)
+
+	_, _, err := cli.ServiceInspectWithRaw(context.Background(), serviceID, types.ServiceInspectOptions{})
```
not sure if we need this one, did you add this for a specific reason? if the service converges, we should know already that it worked?

selansen replied:
again with past experience, we had network creation successful but network inspect failed randomly. So just wanted to make sure service inspect also succeeds. I can remove it if you feel its redundant.

thaJeztah replied:
Can you have a look at this one? (see my earlier comment above https://github.com/moby/moby/pull/38102/files#r233767727)

selansen replied:
missed it I guess. I resolved it now.

selansen commented:
Un related failed test cases:
```----------------------------------------------------------------------
23:24:37 FAIL: docker_api_swarm_test.go:296: DockerSwarmSuite.TestAPISwarmLeaderElection
23:24:37 
23:24:37 [d328cfedec2ed] waiting for daemon to start
23:24:37 [d328cfedec2ed] daemon started
23:24:37 
23:24:37 [d442c79d33aef] waiting for daemon to start
23:24:37 [d442c79d33aef] daemon started
23:24:37 
23:24:37 [daa1ffbbf083c] waiting for daemon to start
23:24:37 [daa1ffbbf083c] daemon started
23:24:37 
23:24:37 [d328cfedec2ed] exiting daemon
23:24:37 Waiting for election to occur...
23:24:37 assertion failed: error is not nil: Error response from daemon: rpc error: code = DeadlineExceeded desc = context deadline exceeded: [d442c79d33aef] (*Daemon).GetNode: NodeInspectWithRaw("5dwrv2ochhvt97z9uoz2xvees") failed```

```FAIL: docker_cli_swarm_test.go:1316: DockerSwarmSuite.TestSwarmClusterRotateUnlockKey
23:55:52 
23:55:52 [d2b319be4eb9a] waiting for daemon to start
23:55:52 [d2b319be4eb9a] daemon started
23:55:52 
23:55:52 [d189849a40655] waiting for daemon to start
23:55:52 [d189849a40655] daemon started
23:55:52 
23:55:52 [d13810b323b91] waiting for daemon to start
23:55:52 [d13810b323b91] daemon started
23:55:52 
23:55:52 [d189849a40655] exiting daemon
23:55:52 [d189849a40655] waiting for daemon to start
23:55:52 [d189849a40655] daemon started
23:55:52 
23:55:52 [d13810b323b91] exiting daemon
23:55:52 [d13810b323b91] waiting for daemon to start
23:55:52 [d13810b323b91] daemon started
23:55:52 
23:55:52 docker_cli_swarm_test.go:1379:
23:55:52     c.Assert(err, checker.IsNil, check.Commentf("%s", outs))
23:55:52 ... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc000994da0), Stderr:[]uint8(nil)} ("exit status 1")
23:55:52 ... Error response from daemon: rpc error: code = Unknown desc = The swarm does not have a leader. It's possible that too few managers are online. Make sure more than half of the managers are online.```

``` --- FAIL: TestClientWithRequestTimeout (0.00s)
22:19:48     client_test.go:251: assertion failed: expected an error, got nil: expected error
22:19:48 time="2018-11-16T22:19:33Z" level=warning msg="Unable to locate plugin: vegetable, retrying in 1s"
22:19:48 time="2018-11-16T22:19:34Z" level=warning msg="Unable to locate plugin: vegetable, retrying in 2s"
22:19:48 time="2018-11-16T22:19:36Z" level=warning msg="Unable to locate plugin: vegetable, retrying in 4s"
22:19:48 time="2018-11-16T22:19:40Z" level=warning msg="Unable to locate plugin: vegetable, retrying in 8s"
22:19:48 FAIL```

thaJeztah commented on api/swagger.yaml:
```diff
@@ -8861,6 +8861,13 @@ paths:
                   nodes in order to reach the containers running on this node. Using this parameter it is possible to
                   separate the container data traffic from the management traffic of the cluster.
                 type: "string"
+              DataPathPort:
```
See my other comment (https://github.com/moby/moby/pull/38102/files#r233649721) looks like the new field in `ClusterInfo` was not yet added to the swagger; can you update the swagger? Here's a diff that you can apply;

```patch
diff --git a/api/swagger.yaml b/api/swagger.yaml
index 5b721ee82c..f12cde5439 100644
--- a/api/swagger.yaml
+++ b/api/swagger.yaml
@@ -2465,6 +2465,16 @@ definitions:
         description: "Whether there is currently a root CA rotation in progress for the swarm"
         type: "boolean"
         example: false
+      DataPathPort:
+        description: |
+          DataPathPort specifies the data path port number for data traffic.
+          Acceptable port range is 1024 to 49151.
+
+          If no port is set or is set to 0, the default port (4789) is used.
+        type: "integer"
+        format: "uint32"
+        default: 4789
+        example: 4789
 
   JoinTokens:
     description: |
```

thaJeztah commented on vendor.conf:
```diff
@@ -37,7 +37,7 @@ github.com/mitchellh/hashstructure 2bca23e0e452137f789efbc8610126fd8b94f73b
 
 # When updating, also update LIBNETWORK_COMMIT in hack/dockerfile/install/proxy.installer accordingly
-github.com/docker/libnetwork 49627167f0585504fd78ed8827529aec57a9618d
+github.com/docker/libnetwork 1f28166bb386cf9223d2d00a28382b0e474be314
```
Can you move the vendoring changes to a separate commit? (this makes it easier to backport if that turns out to be needed)

selansen committed:
VXLAN UDP Port configuration support
This commit contains changes to configure DataPathPort
option. By default we use 4789 port number. But this commit
will allow user to configure port number during swarm init.
DataPathPort can't be modified after swarm init.
Signed-off-by: selansen <elango.siva@docker.com>

selansen committed:
Vendor commit for  VXLAN UDP Port configuration support
This commit brings Swarmkit and Libnetwork library changes
Signed-off-by: selansen <elango.siva@docker.com>

selansen commented:
@thaJeztah PTAL.

thaJeztah commented on integration/network/service_test.go:
```diff
@@ -321,6 +321,69 @@ func noServices(client client.ServiceAPIClient) func(log poll.LogT) poll.Result
+
+	poll.WaitOn(t, serviceRunningCount(cli, serviceID, instances), swarm.ServicePoll)
+
+	_, _, err = cli.ServiceInspectWithRaw(context.Background(), serviceID, types.ServiceInspectOptions{})
```
Same for this one (see https://github.com/moby/moby/pull/38102/files#r233767727)

thaJeztah reviewed (changes requested):
We're getting close; can you look at the `ServiceInspectWithRaw` comments that I left earlier?

Let me try to rattle some people to get the remaining SwarmKit change merged

selansen commented:
let me push swarmkit proto change too. I thought you wanted to bring in separate PR.

tiborvass commented:
@selansen you can now update swarmkit in vendor.conf to use 8af8c420f491f006ab1730e08d446a795b1667d7 which contains the fix from @thaJeztah

tiborvass commented:
@selansen vendor.conf conflict!

selansen commented:
Will look into it

On Wed, Nov 21, 2018 at 7:18 PM Tibor Vass <notifications@github.com> wrote:

> @selansen <https://github.com/selansen> vendor.conf conflict!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/moby/moby/pull/38102#issuecomment-440860134>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKEKn0gDpA5tPe_6eJ0OUxrDuoIaYyJ2ks5uxe1LgaJpZM4YAqNb>
> .
>

selansen commented:
@thaJeztah @tiborvass PTAL

thaJeztah commented on daemon/cluster/listen_addr.go:
```diff
@@ -123,6 +123,25 @@ func validateDefaultAddrPool(defaultAddrPool []string, size uint32) error {
+func getDataPathPort(portNum uint32) (uint32, error) {
+	// if the value comes as 0 by any reason we set it to default value 4789
+	if portNum == 0 {
+		portNum = 4789
```
~Did something go wrong with a rebase? I thought you added a `const` for the default port 🤔~

edit: ah; never mind, that was in libnetwork https://github.com/docker/libnetwork/blob/87b1efc194575f94618d1185e808c4794f3d1c3d/drivers/overlay/overlayutils/utils.go#L14-L27

thaJeztah commented on internal/test/daemon/swarm.go:
```diff
@@ -85,6 +85,7 @@ func (d *Daemon) SwarmInit(t assert.TestingT, req swarm.InitRequest) {
 		req.DefaultAddrPool = d.DefaultAddrPool
 		req.SubnetSize = d.SubnetSize
 	}
+	req.DataPathPort = d.DataPathPort
```
Didn't notice this change really, why was this needed? If no port is specified, won't SwarmKit / libnetwork still set the default automatically? (if not, that would be a problem, because older clients won't set this value).

selansen replied:
This is for testing new option DataPathPort. You are right,  if  port is not specified , libnetwork will use default value. if no port is specified , here the value ill be 0. also this test is going to run only on API version 40 and above.

thaJeztah replied:
But this is setting the port now by default, so even if no port is specified in the test, a port is send in the request.

Point is that this utility is used by at least four tests (`TestAPISwarmJoinToken`, `TestUpdateSwarmAddExternalCA`, `TestAPISwarmPromoteDemote`, `TestAPISwarmForceNewCluster`),
and because it's called by `StartAndSwarmInit` (which is called by `AddDaemon` and `NewSwarm`) is used by 130+ other tests, so all of those tests will now change, correct?

thaJeztah replied:
Hm, so I think my confusion here is that we're overriding the `initRequest` with a daemon configuration; looking a bit further;

So, this PR adds a _daemon_ option `DataPathPort` , so that the port gets automatically set when calling `swarm.NewSwarm()`; https://github.com/moby/moby/blob/5d82d77a6df0c95bb5986ace1255de424c700b33/integration/internal/swarm/service.go#L50-L61


I think the issue lies there;

- Unlike (e.g.) `storageDriver`, `DataPathPort` is not a daemon configuration
- The `DataPathPort` must be given when calling `SwarmInit`

So the problem is that `swarm.NewSwarm()` is doing to much for this test; possible resolutions for that;

- Add a `swarmops ...func(*swarm.InitRequest)` to `swarm.NewSwarm()` (so that the options to initialize the swarm can be passed)
- Don't use `swarm.NewSwarm()` in this test, and first create the daemon, start it (`d.StartNode()`), then call `

```go
d := daemon.New(t)
d.StartNode(t)
d.SwarmInit(t, swarm.InitRequest{DataPathPort: 12345})
```

thaJeztah replied:
Looks like we've taken the same approach for `DefaultAddrPool`, which also is not a daemon option (but something to set on `swarm init`).

We should fix that, but let's do that in a follow-up

thaJeztah reviewed (changes requested):
Thanks for updating! One question, but looks good otherwise

thaJeztah reviewed (approved):
LGTM if green, thanks!!

selansen commented:
Failures unrelated to this PR:
FAIL: docker_cli_swarm_test.go:1316: DockerSwarmSuite.TestSwarmClusterRotateUnlockKey

FAIL: docker_cli_swarm_test.go:1128: DockerSwarmSuite.TestSwarmLockUnlockCluster

@tiborvass , @crosbymichael PTAL.

tiborvass reviewed (approved):
LGTM

thaJeztah commented:
All green; ignoring the codecov/patch one

selansen commented:
Thanks @thaJeztah @tiborvass . Let  me update CLI PR n ping you soon.

sgutermann commented:
@selansen first of all, thank you for your work. I am faced with this problem and I just pulled myself a "nightly" which already shows your work.

One maybe strange question on my part ist, what about the docker swarm join command? I am not able to specify the vxlan port here. Does that bother? Will my vxlan magiacally get the right port?

selansen commented:
@sgutermann you don't need to pass vxlan port during swarm join.  swarm init is the only place you will have to pass vxlan port number.

During swarm join event , swarm manager will propagate vxlan information to the joining node. It's all taken care so user doesn't need to remember each time when they invoke swarm join.
